src/decorators/legacy/decorator-utils.ts

```diff
@@ -4,41 +4,121 @@ export type PropertyOrGetterDescriptor = (
   propertyDescriptor?: PropertyDescriptor,
 ) => void;
 
-export interface ParameterizedPropertyDecorator<T> {
+export interface OptionalParameterizedPropertyDecorator<T> {
+  // @Decorator()
+  (): PropertyOrGetterDescriptor;
+  // @Decorator(value)
   (options: T): PropertyOrGetterDescriptor;
-
+  // @Decorator
   (target: Object, propertyName: string | symbol, propertyDescriptor?: PropertyDescriptor): void;
 }
 
+export interface RequiredParameterizedPropertyDecorator<T> {
+  // @Decorator(value)
+  (options: T): PropertyOrGetterDescriptor;
+}
+
+export const DECORATOR_NO_DEFAULT = Symbol('DECORATOR_NO_DEFAULT');
+
+/**
+ * Creates a decorator that MUST receive a parameter
+ *
+ * @param name
+ * @param callback The callback that will be executed once the decorator is applied.
+ */
+export function createParameterizedPropertyDecorator<T>(
+  name: string,
+  callback: (
+    option: T,
+    target: Object,
+    propertyName: string | symbol,
+    propertyDescriptor: PropertyDescriptor | undefined,
+  ) => void,
+): RequiredParameterizedPropertyDecorator<T> {
+  return createOptionallyParameterizedPropertyDecorator(name, DECORATOR_NO_DEFAULT, callback);
+}
+
 /**
- * Makes a decorator that can optionally receive a parameter
+ * Creates a decorator that can optionally receive a parameter
  *
+ * @param name
  * @param defaultValue The value to use if no parameter is provided.
  * @param callback The callback that will be executed once the decorator is applied.
  */
-export function makeParameterizedPropertyDecorator<T>(
-  defaultValue: T,
+export function createOptionallyParameterizedPropertyDecorator<T>(
+  name: string,
+  defaultValue: T | typeof DECORATOR_NO_DEFAULT,
   callback: (
     option: T,
     target: Object,
     propertyName: string | symbol,
     propertyDescriptor: PropertyDescriptor | undefined,
   ) => void,
-): ParameterizedPropertyDecorator<T> {
-  return function decorator(...args: [options: T] | Parameters<PropertyOrGetterDescriptor>) {
-    if (args.length === 1) {
+): OptionalParameterizedPropertyDecorator<T> {
+  return function decorator(...args: [] | [options: T] | Parameters<PropertyOrGetterDescriptor>) {
+    // note: cannot use <= 1, because TypeScript uses this to infer the type of "args".
+    if (args.length === 0 || args.length === 1) {
```

Comment:
These signatures can get a bit unwieldy, we may want to create some sort of DecoratorOptionsDTO or something to clear up a lot of these methods

Reply:
Decorators are so annoying to implement, I'm so annoyed they didn't take the years of feedback from the legacy decorators into account and still went with the same super annoying design of not auto-inserting parentheses (https://github.com/tc39/proposal-decorators/issues/81#issuecomment-713167505)

Anyway, what do you have in mind to clean this up?